### PR TITLE
package log as a lib

### PR DESCRIPTION
**Pull request by lcnascimento**

### CODE REVIEW

File: lib/infra/log/log.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	LevelCritical Level = "CRITICAL"
+)
+
+var levelPriority = map[Level]int{
+	LevelCritical: 5,
+	LevelError:    10,
+	LevelWarning:  15,
+	LevelInfo:     20,
+	LevelDebug:    25,
+}
```
Comment (RaiTamarindo) [inline lib/infra/log/log.go:31]: I got this idea to remove repeated literals from the code:
```suggestion
// Level indicates the severity of the data being logged
type Level int

const (
	// LevelDebug are debug or trace information
	LevelDebug Level = iota
	// LevelInfo are routine information
	LevelInfo
	// LevelWarning warns about events the might cause problems to the system
	LevelWarning
	// LevelError alerts about events that are likely to cause problems
	LevelError
	// LevelCritical alerts about severe problems. Most of the time, needs some human intervention ASAP
	LevelCritical
)

func (l Level) String() string {
	return []string{
		"DEBUG",
		"INFO",
		"WARNING",
		"ERROR",
		"CRITICAL",
	}[l]
}
```

File: lib/infra/log/log.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	if _, ok := levelPriority[in.Level]; !ok {
+		in.Level = LevelInfo
+	}
+
+	return &Logger{level: in.Level, attributes: in.Attributes}
+}
+
+// Debug logs debug data
+func (l Logger) Debug(ctx context.Context, msg string, args ...interface{}) {
+	if levelPriority[l.level] >= 25 {
```
Comment (RaiTamarindo) [inline lib/infra/log/log.go:62]: Thus, you can do this:
```suggestion
	if l.level >= LevelDebug {
```

File: lib/infra/log/log.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+// Critical logs critical data
+func (l Logger) Critical(ctx context.Context, err error) {
+	if levelPriority[l.level] >= 5 {
+		l.print(ctx, err.Error(), LevelCritical)
+	}
+}
+
+type logData struct {
+	Level      Level                        `json:"level"`
```
Comment (RaiTamarindo) [inline lib/infra/log/log.go:96]:
```suggestion
	Level      string                        `json:"level"`
```

File: lib/infra/log/log.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	Message    string                       `json:"message"`
+	Attributes map[LogAttribute]interface{} `json:"attributes,omitempty"`
+}
+
+func (l Logger) print(ctx context.Context, msg string, level Level) {
+	data, _ := json.Marshal(logData{
+		Level:      level,
```
Comment (RaiTamarindo) [inline lib/infra/log/log.go:103]:
```suggestion
		Level:      level.String(),
```

File: lib/infra/log/log.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// Critical logs critical data
+func (l Logger) Critical(ctx context.Context, err error) {
+	if levelPriority[l.level] >= 5 {
+		l.print(ctx, err.Error(), LevelCritical)
+	}
+}
+
+type logData struct {
```
Comment (RaiTamarindo) [inline lib/infra/log/log.go:95]: I think we can add a `timestamp` attribute to it
```suggestion
type logData struct {
	Timestamp      string                        `json:"timestamp"`
```

**Review by RaiTamarindo — CHANGES_REQUESTED** [review]

File: lib/infra/log/log.go (hunk 1 above)
Comment (lcnascimento) [reply lib/infra/log/log.go]: @RaiTamarindo nice call!!

**Pull request merged**